### PR TITLE
add the popover component

### PR DESCRIPTION
**Pull request by libangrui**

### CODE REVIEW

File: src/components/Base/Popover/index.scss [hunk 1]
```diff
@@ -0,0 +1,115 @@
+@import '~scss/variables';
+
+.popover {
+  :global {
```
Comment (iwisunny) [inline src/components/Base/Popover/index.scss]: `:global` should be in top-level block

File: src/components/Base/Popover/index.jsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    placement: 'bottom',
+    disabled: false,
+  };
+
+  getOverlay = () => {
+    const { prefixCls, title, content } = this.props;
+
+    return (
+      <div>
+        {title && <div className={`${prefixCls}-title`}>{title}</div>}
```
Comment (iwisunny) [inline src/components/Base/Popover/index.jsx]: if `prefixCls` is empty string, `${prefixCls}-title` is `-title`, is it right ?

File: src/components/Base/Popover/index.scss [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            border-width: 7px;
+            border-color: #fff transparent transparent transparent;
+            z-index: 2;
+          }
+        }
+      }
+      &[data-placement^="bottom"] {
+        margin-top: 8px;
+    
+        .pi-popover-popper_arrow {
```
Comment (iwisunny) [inline src/components/Base/Popover/index.scss]: `pi-popover-popper_arrow` seems repeated, can move it to a base section then extend it, or abstract it to a function

File: src/components/Base/Tooltip/index.jsx [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    placement: 'top',
+    disabled: false,
+    visible: false,
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      visible: props.visible,
```
Comment (iwisunny) [inline src/components/Base/Tooltip/index.jsx]: shoud not mix props with state

File: src/components/Base/Tooltip/index.jsx [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    };
+    this.disabled = props.disabled;
```
Comment (iwisunny) [inline src/components/Base/Tooltip/index.jsx]: no need to assign `this.disabled`

File: src/components/Base/Tooltip/index.jsx [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    const placementMapper = {
```
Comment (iwisunny) [inline src/components/Base/Tooltip/index.jsx]: i think this mapper is redundent, can just use `this.props.placement` to control

File: src/components/Base/Tooltip/index.jsx [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+      left: 'left',
+      right: 'right',
+      top: 'top',
+      bottom: 'bottom',
+    };
+    this.placement = placementMapper[this.props.placement] || 'bottom';
+
+    const triggerMapper = {
```
Comment (iwisunny) [inline src/components/Base/Tooltip/index.jsx]: same problem as above

File: src/components/Base/Tooltip/index.jsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    const { prefixCls, className, content, children } = this.props;
+
+    return (
+      <Manager
+        className={classNames(styles.tooltip, {
+          [className]: className,
+        })}
+      >
+        <Target
+          innerRef={c => { this.target = findDOMNode(c); }}
```
Comment (iwisunny) [inline src/components/Base/Tooltip/index.jsx]: c is the DOM node, no need to call findDOMNode ?

File: src/components/Base/Tooltip/index.jsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        >
+          {children}
+        </Target>
+        <Popper
+          className={classNames(styles.popper, `${prefixCls}-popper`, {
+            [styles.active]: this.state.visible,
```
Comment (iwisunny) [inline src/components/Base/Tooltip/index.jsx]: `this` scope is changed? maybe should hoist to function top-level

Comment (libangrui) [reply src/components/Base/Tooltip/index.jsx]: ok

File: src/components/Base/Tooltip/index.jsx (hunk 8 above)
Comment (libangrui) [reply src/components/Base/Tooltip/index.jsx]: yes

File: src/components/Base/Tooltip/index.jsx (hunk 4 above)
Comment (libangrui) [reply src/components/Base/Tooltip/index.jsx]: uh... why?

File: src/components/Base/Popover/index.scss (hunk 1 above)
Comment (libangrui) [reply src/components/Base/Popover/index.scss]: It is worth while to discuss the problem. @leoendless

File: src/components/Base/Popover/index.jsx (hunk 2 above)
Comment (leoendless) [reply src/components/Base/Popover/index.jsx]: prefixCls will use defaultProps value if no value parsed.

File: src/components/Base/Popover/index.scss (hunk 1 above)
Comment (leoendless) [reply src/components/Base/Popover/index.scss]: ``:global`` under ``.popover`` will only effect elements under local style ``.popover``

File: src/components/Base/Popover/index.jsx (hunk 2 above)
Comment (iwisunny) [reply src/components/Base/Popover/index.jsx]: i know, i mean, defaultProp of `prefixCls` could be `pi-popover-`, `${prefixCls}-title` change as `${prefixCls}title` is better, so when `prefixCls` is `''`, that concat string will not prefixed with a `-`

File: src/components/Base/Popover/index.scss (hunk 1 above)
Comment (iwisunny) [reply src/components/Base/Popover/index.scss]: got it

File: src/components/Base/Tooltip/index.jsx (hunk 4 above)
Comment (leoendless) [reply src/components/Base/Tooltip/index.jsx]: if you want to control visibility by props, an ``onVisibleChange `` callback is also needed

File: src/components/Base/Tooltip/index.jsx [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+  showPopper = () => {
+    this.setState({ visible: true });
+  }
+
+  hidePopper = () => {
+    this.setState({ visible: false });
+  }
+
+  handleTogglePopper = () => {
+    this.state.isShow ? this.hidePopper() : this.showPopper();
```
Comment (leoendless) [inline src/components/Base/Tooltip/index.jsx]: where is the ``this.state.isShow`` assigned

File: src/components/Base/Tooltip/index.jsx [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+      target.removeEventListener('click', this.handleTogglePopper);
+      document.removeEventListener('click', this.handleDocumentClick);
+    } else {
+      target.removeEventListener('mouseenter', this.showPopper);
+      target.removeEventListener('mouseleave', this.hidePopper);
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if ('visible' in nextProps) {
```
Comment (leoendless) [inline src/components/Base/Tooltip/index.jsx]: add ``nextProps.visible !== this.state.visible`` seems better

File: src/components/Base/Tooltip/index.jsx (hunk 4 above)
Comment (iwisunny) [reply src/components/Base/Tooltip/index.jsx]: sorry, i forgot the `componentWillReceiveProps ` :)

File: src/components/Base/Tooltip/index.scss [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+$bgcolor: $text-color;
+
+.tooltip {
+}
+
+.popper {
+  padding: 4px 8px;
+  border-radius: 3px;
+  font-size: 12px;
```
Comment (leoendless) [inline src/components/Base/Tooltip/index.scss]: ``font-size`` should use variable in ``variables .scss``

File: src/components/Base/Popover/index.jsx (hunk 2 above)
Comment (libangrui) [reply src/components/Base/Popover/index.jsx]: If put `-`, it does not look like `prefixCls`.
When we set this property, `-` is usually not added.

File: src/components/Base/Popover/index.scss (hunk 3 above)
Comment (libangrui) [reply src/components/Base/Popover/index.scss]: The style here has the public part of the independence, not easily abstracted into a method.

**Pull request merged**